In [1]:
#Import pandas, matplotlib.pyplot, and seaborn
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
#import raw data
house_data = pd.read_csv('./raw_data/train_data.csv')

In [3]:
# first 5 row of the dataset
house_data.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

There are missing values in 'Alley', 'PoolQC', Fence, 'MiscFeature' columns in the first 5 rows. 'Id' column is not a predictor in the dataset.

In [4]:
# data types in the data set
house_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

There are 37 columns containing numeric observations and 38 objects columns.

In [5]:
# size of the dataset
house_data.shape

(1460, 81)

The dataset has 1460 rows and 81 columns

Count the missing values in each columns and the percentages 

In [6]:
missing = pd.concat([house_data.isnull().sum(), 100 * house_data.isnull().mean()], axis=1)
missing.columns=['Count', '%']
missing.sort_values(by= 'Count', ascending= False)

Count          %
PoolQC        1453  99.520548
MiscFeature   1406  96.301370
Alley         1369  93.767123
Fence         1179  80.753425
FireplaceQu    690  47.260274
...            ...        ...
ExterQual        0   0.000000
Exterior2nd      0   0.000000
Exterior1st      0   0.000000
RoofMatl         0   0.000000
SalePrice        0   0.000000

[81 rows x 2 columns]

In [16]:
# features that contains about 50% or higher missing values
missing_50 = missing[round(missing['%'], ndigits = -1) > 0]
missing_50.sort_values(by= 'Count', ascending= False)

Count          %
PoolQC         1453  99.520548
MiscFeature    1406  96.301370
Alley          1369  93.767123
Fence          1179  80.753425
FireplaceQu     690  47.260274
LotFrontage     259  17.739726
GarageType       81   5.547945
GarageYrBlt      81   5.547945
GarageFinish     81   5.547945
GarageQual       81   5.547945
GarageCond       81   5.547945

There are 5 columns of 'PoolQC', 'MiscFeature', 'Alley', 'Fence', 'FireplaceQu' which are missing too many values

In [8]:
# drop 'PoolQC', 'MiscFeature', 'Alley', 'Fence', 'FireplaceQu' and 'Id' columns
house_data.drop(columns= ['Id','PoolQC', 'MiscFeature', 'Alley', 'Fence', 'FireplaceQu'], inplace= True)

In [14]:
house_data.shape

(1460, 75)

In [9]:
house_data.head()

MSSubClass MSZoning  LotFrontage  LotArea Street LotShape LandContour  \
0          60       RL         65.0     8450   Pave      Reg         Lvl   
1          20       RL         80.0     9600   Pave      Reg         Lvl   
2          60       RL         68.0    11250   Pave      IR1         Lvl   
3          70       RL         60.0     9550   Pave      IR1         Lvl   
4          60       RL         84.0    14260   Pave      IR1         Lvl   

  Utilities LotConfig LandSlope  ... EnclosedPorch 3SsnPorch ScreenPorch  \
0    AllPub    Inside       Gtl  ...             0         0           0   
1    AllPub       FR2       Gtl  ...             0         0           0   
2    AllPub    Inside       Gtl  ...             0         0           0   
3    AllPub    Corner       Gtl  ...           272         0           0   
4    AllPub       FR2       Gtl  ...             0         0           0   

  PoolArea MiscVal  MoSold  YrSold  SaleType  SaleCondition SalePrice  
0        0       0       2    2008        WD         Normal    208500  
1        0       0       5    2007        WD         Normal    181500  
2        0       0       9    2008        WD         Normal    223500  
3        0       0       2    2006        WD        Abnorml    140000  
4        0       0      12    2008        WD         Normal    250000  

[5 rows x 75 columns]

In [10]:
# missing values in float64 features
house_data.select_dtypes('float64').isnull().sum().sort_values(ascending= False)

LotFrontage    259
GarageYrBlt     81
MasVnrArea       8
dtype: int64

There are some missing values in all three columns that contains float64 data.

In [11]:
# replace missing values with the mean of the remaining observations
house_data['LotFrontage'] = house_data['LotFrontage'].fillna(house_data['LotFrontage'].mean())
house_data['MasVnrArea'] = house_data['MasVnrArea'].fillna(house_data['MasVnrArea'].mean())
house_data['GarageYrBlt'] = house_data['GarageYrBlt'].fillna(house_data['GarageYrBlt'].mean())

In [12]:
# missing values in int64 features
house_data.select_dtypes('int64').isnull().sum().sort_values(ascending= False)

MSSubClass       0
OpenPorchSF      0
KitchenAbvGr     0
TotRmsAbvGrd     0
Fireplaces       0
GarageCars       0
GarageArea       0
WoodDeckSF       0
EnclosedPorch    0
LotArea          0
3SsnPorch        0
ScreenPorch      0
PoolArea         0
MiscVal          0
MoSold           0
YrSold           0
BedroomAbvGr     0
HalfBath         0
FullBath         0
BsmtHalfBath     0
BsmtFullBath     0
GrLivArea        0
LowQualFinSF     0
2ndFlrSF         0
1stFlrSF         0
TotalBsmtSF      0
BsmtUnfSF        0
BsmtFinSF2       0
BsmtFinSF1       0
YearRemodAdd     0
YearBuilt        0
OverallCond      0
OverallQual      0
SalePrice        0
dtype: int64

No missing value in int64 features

In [13]:
house_data.select_dtypes(include= 'object')

MSZoning Street LotShape LandContour Utilities LotConfig LandSlope  \
0          RL   Pave      Reg         Lvl    AllPub    Inside       Gtl   
1          RL   Pave      Reg         Lvl    AllPub       FR2       Gtl   
2          RL   Pave      IR1         Lvl    AllPub    Inside       Gtl   
3          RL   Pave      IR1         Lvl    AllPub    Corner       Gtl   
4          RL   Pave      IR1         Lvl    AllPub       FR2       Gtl   
...       ...    ...      ...         ...       ...       ...       ...   
1455       RL   Pave      Reg         Lvl    AllPub    Inside       Gtl   
1456       RL   Pave      Reg         Lvl    AllPub    Inside       Gtl   
1457       RL   Pave      Reg         Lvl    AllPub    Inside       Gtl   
1458       RL   Pave      Reg         Lvl    AllPub    Inside       Gtl   
1459       RL   Pave      Reg         Lvl    AllPub    Inside       Gtl   

     Neighborhood Condition1 Condition2  ... Electrical KitchenQual  \
0         CollgCr       Norm       Norm  ...      SBrkr          Gd   
1         Veenker      Feedr       Norm  ...      SBrkr          TA   
2         CollgCr       Norm       Norm  ...      SBrkr          Gd   
3         Crawfor       Norm       Norm  ...      SBrkr          Gd   
4         NoRidge       Norm       Norm  ...      SBrkr          Gd   
...           ...        ...        ...  ...        ...         ...   
1455      Gilbert       Norm       Norm  ...      SBrkr          TA   
1456       NWAmes       Norm       Norm  ...      SBrkr          TA   
1457      Crawfor       Norm       Norm  ...      SBrkr          Gd   
1458        NAmes       Norm       Norm  ...      FuseA          Gd   
1459      Edwards       Norm       Norm  ...      SBrkr          TA   

     Functional GarageType GarageFinish GarageQual GarageCond PavedDrive  \
0           Typ     Attchd          RFn         TA         TA          Y   
1           Typ     Attchd          RFn         TA         TA          Y   
2           Typ     Attchd          RFn         TA         TA          Y   
3           Typ     Detchd          Unf         TA         TA          Y   
4           Typ     Attchd          RFn         TA         TA          Y   
...         ...        ...          ...        ...        ...        ...   
1455        Typ     Attchd          RFn         TA         TA          Y   
1456       Min1     Attchd          Unf         TA         TA          Y   
1457        Typ     Attchd          RFn         TA         TA          Y   
1458        Typ     Attchd          Unf         TA         TA          Y   
1459        Typ     Attchd          Fin         TA         TA          Y   

     SaleType SaleCondition  
0          WD        Normal  
1          WD        Normal  
2          WD        Normal  
3          WD       Abnorml  
4          WD        Normal  
...       ...           ...  
1455       WD        Normal  
1456       WD        Normal  
1457       WD        Normal  
1458       WD        Normal  
1459       WD        Normal  

[1460 rows x 38 columns]